src/components/tl/ucp/allgather/allgather_bruck.c

```diff
@@ -0,0 +1,254 @@
+    return status;
+}
+
+ucc_status_t ucc_tl_ucp_allgather_bruck_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_status_t       global_status = UCC_OK;
+    ucc_status_t       status;
+
+    tl_trace(UCC_TASK_LIB(task), "ucc_tl_ucp_allgather_bruck_finalize");
+
+    if (task->allgather_bruck.scratch_header != NULL) {
+        /* deallocate scratch buffer */
+        global_status = ucc_mc_free(task->allgather_bruck.scratch_header);
+        if (ucc_unlikely(global_status != UCC_OK)) {
+            tl_error(UCC_TASK_LIB(task),
+                     "failed to free scratch buffer memory");
+        }
+        task->allgather_bruck.scratch_size = 0;
+    }
+
+    status = ucc_tl_ucp_coll_finalize(&task->super);
+    if (ucc_unlikely(status != UCC_OK)) {
+        tl_error(UCC_TASK_LIB(task),
+                 "failed to finalize allgather bruck collective");
+        global_status = status;
+    }
+    return global_status;
+}
+
+/* Inspired by implementation: https://github.com/open-mpi/ompi/blob/main/ompi/mca/coll/base/coll_base_allgather.c */
+void ucc_tl_ucp_allgather_bruck_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t      *task = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t      *team = TASK_TEAM(task);
+    ucc_rank_t              trank = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t              tsize = UCC_TL_TEAM_SIZE(team);
+    void                   *rbuf  = TASK_ARGS(task).dst.info.buffer;
+    ucc_memory_type_t       rmem  = TASK_ARGS(task).dst.info.mem_type;
+    ucc_datatype_t          dt    = TASK_ARGS(task).dst.info.datatype;
+    size_t                  count = TASK_ARGS(task).dst.info.count;
+    ucc_mc_buffer_header_t *scratch_header =
+        task->allgather_bruck.scratch_header;
+    size_t       scratch_size = task->allgather_bruck.scratch_size;
+    size_t       data_size    = (count / tsize) * ucc_dt_size(dt);
+    ucc_rank_t   recvfrom, sendto;
+    ucc_status_t status;
+    size_t       blockcount, distance;
+    void        *tmprecv, *tmpsend;
+
+    if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+        return;
+    }
+
+    /* On each step doubles distance */
+    distance = 1 << task->tagged.send_posted;
+    tmpsend  = rbuf;
+    while (distance < tsize) {
+
+        recvfrom = (trank + distance) % tsize;
+        sendto   = (trank + tsize - distance) % tsize;
+
+        tmprecv = PTR_OFFSET(tmpsend, distance * data_size);
+
+        if (distance <= tsize >> 1) {
+            blockcount = distance;
+        } else {
+            /* send-recv all reminder*/
+            blockcount = tsize - distance;
+        }
+
+        /* Sendreceive */
+        UCPCHECK_GOTO(ucc_tl_ucp_send_nb(tmpsend, blockcount * data_size, rmem,
+                                         sendto, team, task),
+                      task, out);
+        UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(tmprecv, blockcount * data_size, rmem,
```

Comment:
is it possible to recv some blocks into scratch buffer and some into actual dst buffer so that we can save memory  copy in post processing step?

Reply:
hm, not sure, anyway, it would increase complexity and reduce readability of code. Need to think or continue as part of optimization work.